src/state_circuit/state.rs

```diff
@@ -0,0 +1,878 @@
+
+        let is_zero = IsZeroChip::configure(
+            meta,
+            |meta| {
+                let padding = meta.query_advice(padding, Rotation::cur());
+                let one = Expression::Constant(F::one());
+                let is_not_padding = one - padding;
+                meta.query_selector(q_not_first) * is_not_padding
+            },
+            |meta| {
+                let address_cur = meta.query_advice(address, Rotation::cur());
+                let address_prev = meta.query_advice(address, Rotation::prev());
+                address_cur - address_prev
+            },
+            address_diff_inv,
+        );
+
+        let monotone = MonotoneChip::<F, ADDRESS_MAX, ADDRESS_INCR, false>::configure(
+            meta,
+            |meta| {
+                let padding = meta.query_advice(padding, Rotation::cur());
+                let one = Expression::Constant(F::one());
+                let is_not_padding = one - padding;
+                meta.query_selector(q_not_first) * is_not_padding
+            },
+            address,
+        );
+
+        let monotone_padding = MonotoneChip::<F, 1, true, false>::configure(
+            meta,
+            |meta| meta.query_selector(q_not_first),
+            padding,
+        );
+
+        // A gate for the first operation (does not need Rotation::prev).
+        meta.create_gate("First row operation", |meta| {
+            let q_first = meta.query_selector(q_first);
+
+            let value = meta.query_advice(value, Rotation::cur());
+            let flag = meta.query_advice(flag, Rotation::cur());
+            let global_counter = meta.query_advice(global_counter, Rotation::cur());
+
+            //      - values[0] == [0]
+            //      - flags[0] == 1
+            //      - global_counters[0] == 0
+
+            // TODO: global_counter can't be 0 for both - memory and stack init rows
+
+            let address = meta.query_advice(address, Rotation::cur());
+            let one = Expression::Constant(F::one());
+            let max_address = Expression::Constant(F::from_u64(ADDRESS_MAX as u64));
+            // For stack init operation: ADDRESS_INCR is false (thus 0) and the address must be ADDRESS_MAX (that is 1023)
+            let init_address = (one.clone()
+                - Expression::Constant(F::from_u64(ADDRESS_INCR as u64)))
+                * (max_address - address);
+
+            vec![
+                q_first.clone() * value,
+                q_first.clone() * (one - flag),
+                q_first.clone() * global_counter,
+                q_first * init_address,
+            ]
+        });
+
+        meta.create_gate("State operation", |meta| {
+            let q_not_first = meta.query_selector(q_not_first);
+            // If address_cur != address_prev, this is an `init`. We must constrain:
+            //      - values[0] == [0]
+            //      - flags[0] == 1
+            //      - global_counters[0] == 0
```

Comment:
Why does `global_counter` need to be 0 for a new address? Isn't `global_counter` supposed to keep increasing across all addresses?

Reply:
For a new call context, the `global_counter` should be init with 0.

Reply:
@spartucus This is not a new call context. It just moves to check the next memory/stack address.

Reply:
The `init` at each new address is effectively a "dummy write" of `value = 0`. This handles the case where the first operation at an address is a `read`. Since this address has never been written to, the first `read` should return 0. (This "dummy write" isn't a real operation and doesn't affect the `global_counter`.)

In terms of circuit constraints: note that the `q_read` case  involves querying the value on the previous row `value_prev`. Without a dummy init row, the first read would end up querying the last value from the previous address.

(Thanks @han0110 for explaining this to me.)

Reply:
Thanks @therealyingtong!

I asked @han0110 a couple of related questions related - global_counter can be set to 0 in the init operation as this is not put in the bus mapping. We don't check global_counter uniqueness in the state circuit - this will be done in the evm circuit. In the evm circuit, we will (1) lookup global_counter one by one and (2) check the degree of bus mapping is equal to the number of operations. If a malicious prover skips a write, (1) will fail. If a malicious prover inserts a non-existing write, (2) fails.

@gaswhat: I think global_counter doesn't need to be increasing across all addresses, it only needs to be ordered increasingly for each address.

Reply:
Thanks for the explanation @therealyingtong @miha-stopar. I missed the `init` row at each new address previously.

I have a follow up question. Could we skip this `init` row for each new address to save some rows? It seems to me a bit redundant. I think we just need to prove the first row of a new address must be a `write` operation. Should be enough?

Reply:
@gaswhat I removed stack init rows in #1bdbe5d. This requires some more code and some more gates, but indeed we save quite some rows. I left the memory init rows in there as they seem to be required. I will try to summarize @han0110 explanation below:

Memory init rows are needed to mimic the [EVM interpreter's behaviour](https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L251-L253). For example, when memory is empty and mload(32) is executed (which returns memory[32..64]), EVM expands memory size to 64 and returns 0. Thus, in state circuit we allocate to 0 all addresses that are used to be compatible with EVM. 

We don't have to do such things for stack operations.